pkg/gatewayserver/io/basicstationlns/messages/messages.go

```diff
@@ -0,0 +1,214 @@
+	TypeDownstreamTimeSync                  = "timesync"
+	TypeDownstreamRemoteCommand             = "runcmd"
+	TypeDownstreamRemoteShell               = "rmtsh"
+
+	configHardwareSpecPrefix            = "sx1301"
+	configHardwareSpecNoOfConcentrators = "1"
+)
+
+// DataRates encodes the available datarates of the channel plan for the Station in the format below
+// [0] -> SF (Spreading Factor; Range: 7...12 for LoRa, 0 for FSK)
+// [1] -> BW (Bandwidth; 125/250/500 for LoRa, ignored for FSK)
+// [2] -> DNONLY (Downlink Only; 1 = true, 0 = false)
+type DataRates [16][3]int
+
+// DiscoverQuery contains the unique identifier of the gateway.
+// This message is sent by the gateway.
+type DiscoverQuery struct {
+	EUI basicstation.EUI `json:"router"`
+}
+
+// DiscoverResponse contains the response to the discover query.
+// This message is sent by the Gateway Server.
+type DiscoverResponse struct {
+	EUI   basicstation.EUI `json:"router"`
+	Muxs  basicstation.EUI `json:"muxs,omitempty"`
+	URI   string           `json:"uri,omitempty"`
+	Error string           `json:"error,omitempty"`
+}
+
+// Type returns the message type of the given data.
+func Type(data []byte) (string, error) {
+	msg := struct {
+		Type string `json:"msgtype"`
+	}{}
+	if err := json.Unmarshal(data, &msg); err != nil {
+		return "", err
+	}
+	return msg.Type, nil
+}
+
+// Version contains version information.
+// This message is sent by the gateway.
+type Version struct {
+	Station  string `json:"station"`
+	Firmware string `json:"firmware"`
+	Package  string `json:"package"`
+	Model    string `json:"model"`
+	Protocol int    `json:"protocol"`
+	Features string `json:"features,omitempty"`
+}
+
+// MarshalJSON implements json.Marshaler.
+func (v Version) MarshalJSON() ([]byte, error) {
+	type Alias Version
+	return json.Marshal(struct {
+		Type string `json:"msgtype"`
+		Alias
+	}{
+		Type:  TypeUpstreamVersion,
+		Alias: Alias(v),
+	})
+}
+
+// IsProduction checks the features field for "prod" and returns true if found.
+// This is then used to set debug options in the router config
+func (v Version) IsProduction() bool {
+	if v.Features == "" {
+		return false
+	}
+	if strings.Contains(v.Features, "prod") {
+		return true
+	}
+	return false
+}
+
+// RouterConfig contains the router configuration.
+// This message is sent by the Gateway Server.
+type RouterConfig struct {
+	NetID          []int                   `json:"NetID"`
+	JoinEUI        [][]int                 `json:"JoinEui"`
+	Region         string                  `json:"region"`
+	HardwareSpec   string                  `json:"hwspec"`
+	FrequencyRange []int                   `json:"freq_range"`
+	DataRates      DataRates               `json:"DRs"`
+	SX1301Config   []pfconfig.SX1301Config `json:"sx1301_conf"`
+
+	// These are debug options to be unset in production gateways
+	NoCCA       bool `json:"nocca"`
+	NoDutyCycle bool `json:"nodc"`
+	NoDwellTime bool `json:"nodwell"`
+
+	// TODO: Roundtrip time monitoring
```

Comment:
TODOs are written as `// TODO: Some Description (https://link/to/issue)`
Also, it looks like this is already done, right?

Reply:
FTR; I approved this in other PRs as they were targeting a feature branch. But this should not go to `master` indeed